File: regiondeployer/httputilities.fs
module HttpUtilities

open System
open System.Collections.Generic
open System.Linq
open System.Net.Http
open System.Text
open System.Threading.Tasks



let basicauthseparator = ":"
let basicauthkey = "Authorization"
let dummyurl = Uri("http://localhost")

let getjsoncontent (json: string) =
    new StringContent(json, Encoding.UTF8, "application/json") // Assuming coreconstants.jsonmimetype is "application/json"

let createrequestmessage (url: string) (method: HttpMethod) (content: HttpContent) (headers: KeyValuePair<string, string>[]) =
    let request = new HttpRequestMessage(method, Uri(url))
    request.Content <- content
    headers |> Array.iter (fun kvp -> request.Headers.Add(kvp.Key, kvp.Value))
    request

let Clone (req: HttpRequestMessage) =
    let clone = new HttpRequestMessage(req.Method, req.RequestUri)
    clone.Content <- req.Content
    clone.Version <- req.Version
    let sequence = req.Options.AsEnumerable() 
    
    Seq.iter(fun (current : KeyValuePair<string, obj>) -> 
        clone.Options.Append(new KeyValuePair<string, obj>(current.Key, current.Value)) |> ignore)
        sequence
    
    let oldheaders: List<KeyValuePair<string, IEnumerable<string>>> = req.Headers.ToList()
    
    for current in oldheaders do
        clone.Headers.TryAddWithoutValidation(current.Key, current.Value) |> ignore

    clone


let createbasicauthheader (username: string) (password: string) =
    let input = username + basicauthseparator + password
    let bytes = Encoding.UTF8.GetBytes(input)
    let output = Convert.ToBase64String(bytes)
    KeyValuePair<string, string>(basicauthkey, "Basic " + output)

File: regiondeployer/HttpUtilities.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Text;

namespace regiondeployer
{
    public class HttpUtilities
    {
        private readonly string basicAuthSeparator = ":";
        private readonly string basicAuthKey = "Authorization";
        private readonly Uri dummyUrl = new Uri("http://localhost");

        public StringContent GetJsonContent(string json)
        {
            return new StringContent(json, Encoding.UTF8, "application/json");
        }

        public HttpRequestMessage CreateRequestMessage(string url, HttpMethod method, HttpContent content, KeyValuePair<string, string>[] headers)
        {
            var request = new HttpRequestMessage(method, new Uri(url))
            {
                Content = content
            };
            foreach (var kvp in headers)
            {
                request.Headers.Add(kvp.Key, kvp.Value);
            }
            return request;
        }

        public HttpRequestMessage Clone(HttpRequestMessage req)
        {
            var clone = new HttpRequestMessage(req.Method, req.RequestUri)
            {
                Content = req.Content,
                Version = req.Version
            };
            foreach (var option in req.Options.AsEnumerable())
            {
                clone.Options.Append(new KeyValuePair<string, object>(option.Key, option.Value));
            }

            var oldHeaders = req.Headers.ToList();
            foreach (var current in oldHeaders)
            {
                clone.Headers.TryAddWithoutValidation(current.Key, current.Value);
            }

            return clone;
        }

        public KeyValuePair<string, string> CreateBasicAuthHeader(string username, string password)
        {
            var input = username + basicAuthSeparator + password;
            var bytes = Encoding.UTF8.GetBytes(input);
            var output = Convert.ToBase64String(bytes);
            return new KeyValuePair<string, string>(basicAuthKey, "Basic " + output);
        }
    }
}
